File: geo_tuner/core/first_order_fit.py
"""Fit a first-order-plus-delay model to a yaw step response.

The yaw closed loop under the geometric controller is (small angles)

    psi_dot = (1/T) * (psi_ref - psi),   T ~ yawctrl_tau / 2  (+ rate lag)

so a reference step gives  psi(t) = A * (1 - exp(-(t - td)/T)).
Identifying T tells us the *effective* yaw time constant, and

    yawctrl_tau_new = yawctrl_tau_applied * T_target / T_measured

places it at the target (the applied tau and measured T are proportional
through the same unknown efficiency factor, which cancels).
"""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np
from scipy.optimize import least_squares


@dataclass
class FirstOrderFitResult:
    T: float           # time constant [s]
    delay: float       # s
    amplitude: float   # fitted steady state (~ step)
    nrmse: float
    converged: bool
    at_bounds: bool = False

    @property
    def ok(self) -> bool:
        return self.converged and self.nrmse < 0.15 and not self.at_bounds


def first_order_step(t: np.ndarray, T: float) -> np.ndarray:
    t = np.maximum(t, 0.0)
    return np.where(t > 0, 1.0 - np.exp(-t / max(T, 1e-6)), 0.0)


def fit_first_order(t: np.ndarray, y: np.ndarray, step: float,
                    T_guess: float = 0.3,
                    T_bounds: tuple = (0.02, 2.0)) -> FirstOrderFitResult:
    """Fit (T, delay, amplitude); y is the response relative to the
    pre-step value, same sign convention as `step`."""
    t = np.asarray(t, dtype=float)
    y = np.asarray(y, dtype=float)
    if t.size < 10:
        raise ValueError("Need at least 10 samples")
    if abs(step) < 1e-6:
        raise ValueError("step must be nonzero")

    yn = y / step

    def residuals(p):
        T, td, amp = p
        return amp * first_order_step(t - td, T) - yn

    lb = np.array([T_bounds[0], 0.0, 0.3])
    ub = np.array([T_bounds[1], 0.4, 1.7])
    p0 = np.clip(np.array([T_guess, 0.05, 1.0]), lb, ub)
    sol = least_squares(residuals, p0, bounds=(lb, ub), method="trf")
    T, td, amp = sol.x

    def _pinned(v, lo, hi, skip_lo=False):
        span = hi - lo
        return ((not skip_lo and v - lo < 0.01 * span)
                or hi - v < 0.01 * span)

    at_bounds = (_pinned(T, lb[0], ub[0])
                 or _pinned(td, lb[1], ub[1], skip_lo=True)
                 or _pinned(amp, lb[2], ub[2]))

    nrmse = float(np.sqrt(np.mean(sol.fun ** 2)))
    return FirstOrderFitResult(
        T=float(T), delay=float(td), amplitude=float(amp * step),
        nrmse=nrmse, converged=bool(sol.success), at_bounds=bool(at_bounds))

File: src/core/first_order_fit.cpp
#include "geo_tuner/core/first_order_fit.hpp"

#include <algorithm>
#include <cmath>
#include <stdexcept>

#include "geo_tuner/core/least_squares.hpp"

namespace geo_tuner
{
namespace
{

bool pinned(double v, double lo, double hi, bool skip_lo = false)
{
  const double span = hi - lo;
  return (!skip_lo && v - lo < 0.01 * span) || (hi - v < 0.01 * span);
}

}  // namespace

Eigen::VectorXd first_order_step(const Eigen::VectorXd & t_in, double T)
{
  Eigen::VectorXd y(t_in.size());
  const double Tc = std::max(T, 1e-6);
  for (Eigen::Index i = 0; i < t_in.size(); ++i) {
    const double t = std::max(t_in[i], 0.0);
    y[i] = t_in[i] > 0.0 ? 1.0 - std::exp(-t / Tc) : 0.0;
  }
  return y;
}

FirstOrderFitResult fit_first_order(
  const Eigen::VectorXd & t, const Eigen::VectorXd & y, double step,
  double T_guess, std::pair<double, double> T_bounds)
{
  if (t.size() < 10) {
    throw std::invalid_argument("Need at least 10 samples");
  }
  if (std::abs(step) < 1e-6) {
    throw std::invalid_argument("step must be nonzero");
  }

  const Eigen::VectorXd yn = y / step;

  auto residuals = [&t, &yn](const Eigen::VectorXd & p) {
      const double T = p[0], td = p[1], amp = p[2];
      Eigen::VectorXd shifted = t.array() - td;
      return (amp * first_order_step(shifted, T) - yn).eval();
    };

  const Eigen::Vector3d lb(T_bounds.first, 0.0, 0.3);
  const Eigen::Vector3d ub(T_bounds.second, 0.4, 1.7);
  const Eigen::Vector3d p0(T_guess, 0.05, 1.0);
  const auto sol = least_squares_bounded(residuals, clip(p0, lb, ub), lb, ub);

  const double T = sol.x[0], td = sol.x[1], amp = sol.x[2];
  const bool at_bounds =
    pinned(T, lb[0], ub[0]) ||
    pinned(td, lb[1], ub[1], /*skip_lo=*/true) ||
    pinned(amp, lb[2], ub[2]);

  FirstOrderFitResult out;
  out.T = T;
  out.delay = td;
  out.amplitude = amp * step;
  out.nrmse = std::sqrt(sol.fun.squaredNorm() / static_cast<double>(sol.fun.size()));
  out.converged = sol.success;
  out.at_bounds = at_bounds;
  return out;
}

}  // namespace geo_tuner
